server/datastore/mysql/hosts.go

```diff
@@ -63,6 +63,25 @@ func (d *Datastore) NewHost(ctx context.Context, host *fleet.Host) (*fleet.Host,
 }
 
 func (d *Datastore) SaveHost(ctx context.Context, host *fleet.Host) error {
+	errCh := make(chan error)
+	select {
+	case <-ctx.Done():
+		return nil
+	case d.writeCh <- itemToWrite{
+		ctx:   ctx,
+		errCh: errCh,
+		item:  host,
+	}:
+		select {
+		case <-ctx.Done():
```

Comment:
If for some reason `<-ctx.Done()`, then the write to the unbuffered `errCh` on `writeChanLoop` could block forever...

Reply:
With the `defer close()` it should be better.

Reply:
Mhm, but now `writeChanLoop` may panic when writing to `errCh`.